skyportal/handlers/api/candidate.py

```diff
@@ -0,0 +1,280 @@
+            required: True
+            schema:
+              type: string
+        requestBody:
+          content:
+            application/json:
+              schema: SourceNoID
+        responses:
+          200:
+            content:
+              application/json:
+                schema: Success
+          400:
+            content:
+              application/json:
+                schema: Error
+        """
+        # Ensure user has access to candidate
+        c = Source.get_if_owned_by(candidate_id, self.current_user,
+                                   groups_attr="candidate_groups")
+        if c is None:
+            return self.error("Invalid ID or inadequate permssions.")
+        data = self.get_json()
+        data["id"] = candidate_id
+
+        save_as_source = data.pop("saveAsSource", False)
+        if save_as_source:
+            group_ids = data.get("groupIDs", None)
+            if group_ids is None:
+                return self.error("Required groupIDs field missing.")
+            groups = Group.query.filter(Group.id.in_(group_ids)).all()
+            if len(groups) == 0:
+                return self.error("Invalid group ID(s).")
+            c.is_source = True
+            c.source_groups = groups
+            for g in groups:
+                gs = GroupSource.query.filter(GroupSource.group_id == g.id).filter(
+                    GroupSource.source_id == c.id
+                ).first()
+                if gs is None:
+                    gs = GroupSource(group_id=g.id, source_id=c.id,
+                                     saved_as_source_by_id=self.current_user.id,
+                                     saved_as_source_at_time=arrow.now())
+                    DBSession.add(gs)
+                else:
+                    gs.saved_as_source_by_id = self.current_user.id
+                    gs.saved_as_source_at_time = arrow.now()
+            DBSession.commit()
+            self.push_all("skyportal/FETCH_SOURCES")
+            return self.success(
+                data={"candidate": c}, action="skyportal/FETCH_CANDIDATES"
+            )
+
+        schema = Source.__schema__()
+        try:
+            schema.load(data)
+        except ValidationError as e:
+            return self.error(
+                "Invalid/missing parameters: " f"{e.normalized_messages()}"
+            )
+        DBSession().commit()
+
+        self.push_all(action="skyportal/FETCH_CANDIDATES")
+        return self.success()
+
+    @permissions(["Manage sources"])
+    def delete(self, candidate_id):
+        """
+        ---
+        description: Delete a candidate
+        parameters:
+          - in: path
+            name: candidate_id
+            required: true
+            schema:
+              type: string
+        responses:
+          200:
+            content:
+              application/json:
+                schema: Success
+        """
+        _ = Source.get_if_owned_by(candidate_id, self.current_user,
```

Comment:
Will this raise an exception if not owned? Perhaps we need a `is_owned_by` which returns a bool.

Reply:
If not owned, the request will return with an error code and message of `"Inadequate permissions."`